### PR TITLE
Refactor Evasion Detectors

### LINKED ISSUE #1913: Refactor Evasion Detectors
**Is your feature request related to a problem? Please describe.**
The current implementation of evasion detectors uses the `BinaryInputDetector` and `BinaryActivationDetector` as the base (`art.defenses.evasion.detector`) which limits future implementation of other other detectors such as Blacklight.

**Describe the solution you'd like**
Refactor `art.defenses.detector.evasion` such that `detector.py` becomes an abstract class and the current implementation is extracted out as `binary_detector.py` which inherits from this new abstract class. This follows the same style and structure of other modules in ART and allows future implementations of evasion detectors.

**Describe alternatives you've considered**
N/A

**Additional context**
This change will be followed with implementations for the L-p norm and Blacklight detector.

### PR DESCRIPTION
**Pull request by f4str**
# Description

Refactor the `art.defences.detector.evasion` module by creating a base abstract class `EvasionDetector` which all of the evasion detectors now inherit from. This base class has a `.detect()` method which will return a report and binary numpy array for which samples are detected adversarial. The `BinaryInputDetector`, `BinaryActivationDetector`, and `SubsetScanningDetector` classes have been refactored accordingly to inherit from `EvasionDetector` and use the `.detect()` method.

The unit tests have been updated accordingly and now use the pytest framework, so they have been removed from legacy tests.

This is technically a "breaking change" since the module has been refactored, but existing functionality is unchanged. Documentation has been updated along with updating `notebooks/detection_adversarial_samples_cifar10.ipynb` to use the new evasion detector structure.

Fixes #1913 

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

- [x] Unit tests for `BinaryInputDetector` using Keras, TensorFlow, and PyTorch
- [x] Unit tests for `BinaryActivationDetector` using Keras, TensorFlow, and PyTorch
- [x] Unit tests for `SubsetScanningDetector` using Keras, TensorFlow, and PyTorch

**Test Configuration**:
- OS
- Python version
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/defences/detector/evasion/binary_activation_detector.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    Binary detector of adversarial samples coming from evasion attacks. The detector uses an architecture provided by
+    the user and is trained on the values of the activations of a classifier at a given layer.
+    """
+
+    defence_params = ["classifier", "detector", "layer"]
+
+    def __init__(
+        self,
+        classifier: "ClassifierNeuralNetwork",
+        detector: "ClassifierNeuralNetwork",
```
Comment (beat-buesser) [inline art/defences/detector/evasion/binary_activation_detector.py:48]: Let's use `CLASSIFIER_NEURALNETWORK_TYPE` from `art.utils` which facilitates type checking:
```suggestion
        classifier: CLASSIFIER_NEURALNETWORK_TYPE,
        detector: CLASSIFIER_NEURALNETWORK_TYPE,
```

File: art/defences/detector/evasion/binary_input_detector.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+class BinaryInputDetector(EvasionDetector):
+    """
+    Binary detector of adversarial samples coming from evasion attacks. The detector uses an architecture provided by
+    the user and trains it on data labeled as clean (label 0) or adversarial (label 1).
+    """
+
+    defence_params = ["detector"]
+
+    def __init__(self, detector: "ClassifierNeuralNetwork") -> None:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/binary_input_detector.py:45]:
```suggestion
    def __init__(self, detector: CLASSIFIER_NEURALNETWORK_TYPE) -> None:
```

File: art/defences/detector/evasion/binary_input_detector.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+import logging
+from typing import Tuple, TYPE_CHECKING
+
+import numpy as np
+
+from art.defences.detector.evasion.evasion_detector import EvasionDetector
+
+if TYPE_CHECKING:
+    from art.estimators.classification.classifier import ClassifierNeuralNetwork
```
Comment (beat-buesser) [inline art/defences/detector/evasion/binary_input_detector.py:32]:
```suggestion
    from art.utils import CLASSIFIER_NEURALNETWORK_TYPE
```

File: art/defences/detector/evasion/evasion_detector.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    def get_params(self) -> Dict[str, Any]:
+        """
+        Returns dictionary of parameters used to run defence.
+
+        :return: Dictionary of parameters of the method.
+        """
+        dictionary = {param: getattr(self, param) for param in self.defence_params}
+        return dictionary
+
+    def _check_params(self) -> None:
```
Comment (beat-buesser) [inline art/defences/detector/evasion/evasion_detector.py:89]: Should we implement `_check_params` and apply it in `__init__` of the class implementations?

File: docs/modules/defences/detector_evasion.rst [hunk 5]
```diff
@@ -2,6 +2,11 @@
 ====================================
 .. automodule:: art.defences.detector.evasion
 
+Base Class
```
Comment (beat-buesser) [inline docs/modules/defences/detector_evasion.rst:5]: Have you tested rebuilding the docs?

File: run_tests.sh [hunk 6]
```diff
@@ -143,8 +143,6 @@ else
                          "tests/defences/test_rounded.py" \
                          "tests/defences/test_thermometer_encoding.py" \
                          "tests/defences/test_variance_minimization.py" \
-                         "tests/defences/detector/evasion/subsetscanning/test_detector.py" \
```
Comment (beat-buesser) [inline run_tests.sh:146]: It looks like the new tests are not yet being run. Please change lines 22 and 23 of this file to:
```python
pytest --cov-report=xml --cov=art --cov-append  -q -vv tests/defences/detector/evasion --framework=$framework --durations=0
    if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/detector/evasion tests"; fi
```

File: tests/defences/detector/evasion/test_binary_activation_detector.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_binary_activation_detector.py:1]: Let's use the opportunity to make the tests framework independent. I'll add a few suggestions below.

File: tests/defences/detector/evasion/test_binary_activation_detector.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+from art.attacks.evasion.fast_gradient import FastGradientMethod
+from art.estimators.classification import KerasClassifier, TensorFlowV2Classifier, PyTorchClassifier
+from art.defences.detector.evasion import BinaryActivationDetector
+
+from tests.utils import ARTTestException, get_image_classifier_kr, get_image_classifier_tf, get_image_classifier_pt
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.only_with_platform("keras", "kerastf")
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_binary_activation_detector.py:33]:
```suggestion
```

File: tests/defences/detector/evasion/test_binary_activation_detector.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+def test_binary_activation_detector_keras(art_warning, get_default_mnist_subset):
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_binary_activation_detector.py:34]:
```suggestion
def test_binary_activation_detector(art_warning, get_default_mnist_subset):
```

File: tests/defences/detector/evasion/test_binary_activation_detector.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    from tensorflow.keras import layers, Sequential
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_binary_activation_detector.py:35]:
```suggestion
```

File: tests/defences/detector/evasion/test_binary_activation_detector.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        _, test_detection = detector.detect(x_test)
+        _, test_adv_detection = detector.detect(x_test_adv)
+
+        # Assert there is at least one true positive and negative:
+        nb_true_positives = np.sum(test_adv_detection)
+        nb_true_negatives = len(test_detection) - np.sum(test_detection)
+        assert nb_true_positives > 0
+        assert nb_true_negatives > 0
+    except ARTTestException as e:
+        art_warning(e)
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_binary_activation_detector.py:176]:
```suggestion
```

File: tests/defences/detector/evasion/test_binary_input_detector.py [hunk 12]
```diff
@@ -0,0 +1,183 @@
+# MIT License
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_binary_input_detector.py:1]: We should add similar changes as for test_binary_input_detector.py

File: tests/defences/detector/evasion/test_subsetscanning_detector.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        elif framework == "tensorflow2":
+            classifier, _ = get_image_classifier_tf()
+        elif framework == "pytorch":
+            classifier = get_image_classifier_pt()
+        else:
+            classifier = None
+
+        return classifier
+
+    return _get_classifier
```
Comment (beat-buesser) [inline tests/defences/detector/evasion/test_subsetscanning_detector.py:50]: I think such a fixture should already exist with `image_dl_estimator`.

```suggestion
```

File: docs/modules/defences/detector_evasion.rst (hunk 5 above)
Comment (f4str) [reply docs/modules/defences/detector_evasion.rst:5]: Rebuilding the produces the correct updated docs.

File: art/defences/detector/evasion/evasion_detector.py (hunk 4 above)
Comment (f4str) [reply art/defences/detector/evasion/evasion_detector.py:89]: The `_check_params` method was mainly implemented for future evasion detectors (e.g., blacklight) to use. The three current evasion detectors do not benefit much from using it.

File: docs/modules/defences/detector_evasion.rst (hunk 5 above)
Comment (beat-buesser) [reply docs/modules/defences/detector_evasion.rst:5]: Thank you!

**Review by beat-buesser — APPROVED** [review]
Hi @f4str Thank you very much for refactoring the evasion detectors, this should be very useful for many users! I also like how you made the detector tests framework independent!

**Pull request merged**